File: facade/src/main/java/org/tdf/sunflower/consensus/AbstractValidator.java
package org.tdf.sunflower.consensus;

import lombok.NonNull;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.ByteUtil;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.facade.RepositoryReader;
import org.tdf.sunflower.facade.Validator;
import org.tdf.sunflower.state.Account;
import org.tdf.sunflower.state.StateTrie;
import org.tdf.sunflower.types.*;
import org.tdf.sunflower.vm.Backend;
import org.tdf.sunflower.vm.CallData;
import org.tdf.sunflower.vm.VMExecutor;

import java.util.*;

public abstract class AbstractValidator implements Validator {
    protected final StateTrie<HexBytes, Account> accountTrie;

    public AbstractValidator(StateTrie<HexBytes, Account> accountTrie) {
        this.accountTrie = accountTrie;
    }

    public long getBlockGasLimit() {
        return 0;
    }

    protected BlockValidateResult commonValidate(RepositoryReader rd, @NonNull Block block, @NonNull Block parent) {
        if (block.getBody() == null || block.getBody().isEmpty())
            return BlockValidateResult.fault("missing block body");

        // a block should contains exactly one coin base transaction

        // validate coinbase
        if (!block.getCoinbase().equals(block.getBody().get(0).getReceiveHex())) {
            return BlockValidateResult.fault("block coinbase not equalt to coinbase tx receiver");
        }

        boolean isCoinbase = true;
        for (Transaction t : block.getBody()) {
            // validate transaction signature
            if (!isCoinbase && t.getSignature() == null)
                return BlockValidateResult.fault("unsigned transaction is not allowed here");
            try {
                t.verify();
            } catch (Exception e) {
                return BlockValidateResult.fault(e.getMessage());
            }
            if (!isCoinbase && !t.verifySig()) {
                return BlockValidateResult.fault("verify signature failed");
            }
            isCoinbase = false;
        }
        if (!parent.isParentOf(block) || parent.getHeight() + 1 != block.getHeight()) {
            return BlockValidateResult.fault("dependency is not parent of block");
        }
        if (parent.getCreatedAt() >= block.getCreatedAt()) {
            return BlockValidateResult.fault(
                String.format("invalid timestamp %d at block height %d", block.getCreatedAt(), block.getHeight())
            );
        }
        if (!Transaction.calcTxTrie(block.getBody()).equals(
            block.getTransactionsRoot()
        )) {
            return BlockValidateResult.fault("transactions root not match");
        }
        BlockValidateResult success = BlockValidateResult.success();

        Uint256 totalFee = Uint256.ZERO;
        long gas = 0;

        Map<HexBytes, VMResult> results = new HashMap<>();
        HexBytes currentRoot;
        long currentGas = 0;
        List<TransactionReceipt> receipts = new ArrayList<>();

        try {
            Backend tmp = accountTrie.createBackend(parent.getHeader(), null, false);
            Transaction coinbase = block.getBody().get(0);


            for (Transaction tx : block.getBody().subList(1, block.getBody().size())) {
                if (currentGas > getBlockGasLimit())
                    return BlockValidateResult.fault("block gas overflow");

                VMResult r;

                VMExecutor executor = new VMExecutor(
                    rd,
                    tmp,
                    CallData.fromTransaction(tx, false),
                    Math.min(getBlockGasLimit() - currentGas, tx.getGasLimitAsU256().longValue())
                );
                r = executor.execute();

                results.put(HexBytes.fromBytes(tx.getHash()), r);
                totalFee = totalFee.plus(r.getFee());

                currentGas += r.getGasUsed();
                TransactionReceipt receipt = new TransactionReceipt(
                    ByteUtil.longToBytesNoLeadZeroes(currentGas),
                    new Bloom(),
                    Collections.emptyList()
                );
                receipt.setGasUsed(r.getGasUsed());
                receipt.setExecutionResult(r.getExecutionResult().getBytes());
                currentRoot = tmp.merge();
                receipts.add(receipt);
                tmp = accountTrie.createBackend(parent.getHeader(), currentRoot, null, false);
            }

            tmp.setHeaderCreatedAt(block.getCreatedAt());
            VMExecutor executor = new VMExecutor(rd, tmp, CallData.fromTransaction(coinbase, true), 0);
            VMResult r = executor.execute();
            currentGas += r.getGasUsed();

            TransactionReceipt receipt = new TransactionReceipt(
                ByteUtil.longToBytesNoLeadZeroes(currentGas),
                new Bloom(),
                Collections.emptyList()
            );
            receipt.setGasUsed(r.getGasUsed());
            receipt.setExecutionResult(r.getExecutionResult().getBytes());
            receipts.add(0, receipt);

            HexBytes rootHash = tmp.merge();
            if (!rootHash.equals(block.getStateRoot())) {
                return BlockValidateResult.fault("state trie root not match");
            }

            // validate receipt trie root
            if (!TransactionReceipt.calcReceiptsTrie(receipts).equals(block.getReceiptTrieRoot())) {
                return BlockValidateResult.fault("receipts trie root not match");
            }

        } catch (Exception e) {
            e.printStackTrace();
            return BlockValidateResult.fault("contract evaluation failed or " + e.getMessage());
        }

        List<TransactionInfo> infos = new ArrayList<>();
        for (int i = 0; i < receipts.size(); i++) {
            infos.add(new TransactionInfo(receipts.get(i), block.getHash().getBytes(), i));
        }
        success.setInfos(infos);
        success.setResults(results);
        success.setFee(totalFee);
        success.setGas(gas);
        return success;
    }
}

File: facade/src/main/java/org/tdf/sunflower/consensus/AbstractValidator.kt
package org.tdf.sunflower.consensus

import org.tdf.common.types.Uint256
import org.tdf.common.util.HexBytes
import org.tdf.common.util.hex
import org.tdf.common.util.min
import org.tdf.sunflower.facade.RepositoryReader
import org.tdf.sunflower.facade.TransactionInfo
import org.tdf.sunflower.facade.Validator
import org.tdf.sunflower.state.Account
import org.tdf.sunflower.state.AddrUtil
import org.tdf.sunflower.state.StateTrie
import org.tdf.sunflower.types.*
import org.tdf.sunflower.types.BlockValidateResult.Companion.fault
import org.tdf.sunflower.types.BlockValidateResult.Companion.success
import org.tdf.sunflower.vm.CallContext
import org.tdf.sunflower.vm.CallData
import org.tdf.sunflower.vm.VMExecutor

abstract class AbstractValidator(protected val accountTrie: StateTrie<HexBytes, Account>) : Validator {
    abstract val chainId: Int

    protected fun commonValidate(rd: RepositoryReader, block: Block, parent: Block): BlockValidateResult {
        block.validate().let {
            if (!it.success)
                return fault(it.reason)
        }

        // a block should contains exactly one coin base transaction
        if (block.body.isEmpty()) return fault("missing block body")

        if (block.gasLimit != rd.genesis.gasLimit)
            return fault("invalid block gas limit")

        var isCoinbase = true
        for (t in block.body) {
            // validate transaction signature
            try {
                t.validate()
            } catch (e: Exception) {
                return fault(e.message!!)
            }
            if (!isCoinbase && !t.verifySig)
                return fault("verify signature failed")

            if (!isCoinbase && t.chainId != chainId)
                return fault("invalid chainId ${t.chainId}")

            isCoinbase = false
        }
        if (!parent.isParentOf(block) || parent.height + 1 != block.height) {
            return fault("dependency is not parent of block")
        }
        if (parent.createdAt >= block.createdAt) {
            return fault(String.format("invalid timestamp %d at block height %d", block.createdAt, block.height))
        }
        if (Transaction.calcTxTrie(block.body) != block.transactionsRoot) {
            return fault("transactions root not match")
        }
        var totalFee = Uint256.ZERO
        val results: MutableMap<HexBytes, VMResult> = mutableMapOf()
        val currentBackend = accountTrie.createBackend(parent = parent.header)
        var currentGas: Long = 0
        val receipts: MutableList<TransactionReceipt> = mutableListOf()
        try {
            val coinbase = block.body[0]
            for (tx in block.body.subList(1, block.body.size)) {
                if (currentGas > block.gasLimit) return fault("block gas overflow")
                var r: VMResult
                val executor = VMExecutor.create(
                    rd,
                    currentBackend,
                    CallContext.fromTx(tx, timestamp = block.createdAt, coinbase = block.coinbase, blockHashMap = rd.createBlockHashMap(block.hashPrev).toMap()),
                    CallData.fromTx(tx),
                    (block.gasLimit - currentGas).min(tx.gasLimit)
                )
                r = executor.execute()
                results[tx.hash] = r
                totalFee += r.fee
                currentGas += r.gasUsed
                val receipt = TransactionReceipt(
                    cumulativeGas = currentGas,
                    logInfoList = r.logs,
                    gasUsed = r.gasUsed,
                    result = r.executionResult
                )
                receipts.add(receipt)
            }

            val executor =
                VMExecutor.create(
                    rd,
                    currentBackend,
                    CallContext.fromTx(coinbase, chainId, block.createdAt, block.coinbase, rd.createBlockHashMap(block.hashPrev).toMap()),
                    CallData.fromTx(coinbase),
                    coinbase.gasLimit
                )


            // assert fee
            if (totalFee != coinbase.value) {
                return fault("total fee $totalFee != coinbase value ${coinbase.value}")
            }

            currentBackend.addBalance(AddrUtil.empty(), totalFee)
            val r = executor.execute()

            currentGas += r.gasUsed

            val receipt = TransactionReceipt(
                cumulativeGas = currentGas,
                logInfoList = r.logs,
                gasUsed = r.gasUsed,
                result = r.executionResult
            )

            receipts.add(0, receipt)
            val rootHash = currentBackend.merge()
            if (rootHash != block.stateRoot) {
                return fault("state trie root not match")
            }

            // validate receipt trie root
            if (TransactionReceipt.calcTrie(receipts) != block.receiptsRoot) {
                return fault("receipts trie root not match")
            }

            // validate bloom
            if (TransactionReceipt.bloomOf(receipts).data.hex() != block.logsBloom)
                return fault("logs bloom not match")

        } catch (e: Exception) {
            e.printStackTrace()
            return fault("contract evaluation failed or " + e.message)
        }
        val indices: MutableList<TransactionInfo> = mutableListOf()
        for (i in receipts.indices) {
            indices.add(
                TransactionInfo(TransactionIndex(receipts[i], block.hash, i), block.body[i])
            )
        }
        return success(currentGas, totalFee, results, indices)
    }
}